File: AIDLServiceDemo/app/src/main/java/com/baiheng/aidlservicedemo/MainActivity.java
package com.baiheng.aidlservicedemo;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }
}

/**
 * 抽象主题类
 */
abstract class Subject {
    public  abstract void visit();
}

/**
 * 真实主题类，委托对象
 */
class RealSubject extends Subject {
    @Override
    public void visit() {
        //具体逻辑
    }
}

/**
 * 代理类，代理对象
 */
class ProxySubject extends Subject {

    private RealSubject mSubject;   //持有真实主题类引用

    public ProxySubject(RealSubject realSubject) {
        this.mSubject = realSubject;
    }

    @Override
    public void visit() {
        mSubject.visit();   //调用真实主题类中逻辑方法
    }
}

class Client {
    public static void main(String[] args) {
        //构造真实主题对象，即委托对象
        RealSubject realSubject = new RealSubject();
        //构造代理对象
        ProxySubject proxySubject = new ProxySubject(realSubject);
        //调用代理相关方法
        proxySubject.visit();
    }
}

File: Kotlin/KotlinPracticeFirst/app/src/main/java/com/zhewen/kotlinpracticefirst/MainActivity.kt
package com.zhewen.kotlinpracticefirst

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import dagger.hilt.android.AndroidEntryPoint
import java.lang.Exception
import java.lang.RuntimeException
import javax.inject.Inject
import kotlin.properties.Delegates

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

    @Inject
    lateinit var truck: Truck

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        truck.deliver()
    }
}


class User {
    var name: String by Delegates.observable("初始值") {
            prop, old, new ->
        println("旧值：$old -> 新值：$new")
    }

    fun fail() {
        throw RuntimeException("kkk")
    }
}


//fun main(args: Array<String>) {
//    val user = User()
//    user.name = "第一次赋值"
//    user.name = "第二次赋值"
//}

//调用
 	// User		User2

fun firstQuestion() {

    println("-------firstQuestion------")
    val arr = Array(100000) {
        it + 1
    }
    operationTime {
        println("Array average value is ${arr.average()}")
    }
    val intArray = IntArray(100000) {
        it + 1
    }
    operationTime {
        println("intArray average value is ${intArray.average()}")
    }

    val list = List(100000) { it + 1 }
    operationTime {
        println("List average value is ${list.average()}")
    }
}

fun operationTime(action: (() -> Unit)?) {
    val startTime = System.currentTimeMillis()
    action?.invoke()
    println("time spent on this operation is ${System.currentTimeMillis() - startTime}")
}

fun secondQuestion() {

    println("-------secondQuestion------")
    listOf(21, 40, 11, 33, 78)
        .asSequence()
        .filter {
            it % 3 == 0
        }
        .forEach {
            println(it)
        }
}

fun thirdQuestion(expression: String) {
    println("-------thirdQuestion------")
    val arg = expression.split(" ")
    val operators = mapOf(
        "+" to ::plus,
        "-" to ::minus,
        "*" to ::multiply,
        "/" to ::div
    )
    println("Input:$expression")
    try {
        operators[arg[1]]?.let {
            println("output:${it(arg[0].toInt(), arg[2].toInt())}")
        }
    } catch (e: Exception) {
        println("Wrong expression,please enter the correct expression ${e}")
    }
}

fun plus(arg0: Int, arg1: Int): Int {
    return arg0 + arg1
}

fun minus(arg0: Int, arg1: Int): Int {
    return arg0 - arg1
}

fun multiply(arg0: Int, arg1: Int): Int {
    return arg0 * arg1
}

fun div(arg0: Int, arg1: Int): Int {
    return arg0 / arg1
}
